snap/squashfs/squashfs.go

```diff
@@ -301,8 +304,130 @@ func (s *Snap) ListDir(dirPath string) ([]string, error) {
 	return directoryContents, nil
 }
 
+const maxErrPaths = 10
+
+type errPathsNotReadable struct {
+	paths []string
+}
+
+func (e *errPathsNotReadable) accumulate(p string, fi os.FileInfo) error {
+	if len(e.paths) >= maxErrPaths {
+		return e
+	}
+	if st, ok := fi.Sys().(*syscall.Stat_t); ok {
+		e.paths = append(e.paths, fmt.Sprintf("%s (owner %v:%v mode %#03o)", p, st.Uid, st.Gid, fi.Mode().Perm()))
+	} else {
+		e.paths = append(e.paths, p)
+	}
+	return nil
+}
+
+func (e *errPathsNotReadable) merge(errOther *errPathsNotReadable) error {
+	toTake := maxErrPaths - len(e.paths)
+	if toTake == 0 {
+		return e
+	}
+	if len(errOther.paths) == 0 {
+		return nil
+	}
+	if toTake > len(errOther.paths) {
+		toTake = len(errOther.paths)
+	}
+
+	e.paths = append(e.paths, errOther.paths[0:toTake]...)
+
+	if len(e.paths) == maxErrPaths {
+		return e
+	}
+	return nil
+}
+
+func (e *errPathsNotReadable) asErr() error {
+	if len(e.paths) > 0 {
+		return e
+	}
+	return nil
+}
+
+func (e *errPathsNotReadable) Error() string {
+	var b bytes.Buffer
+
+	b.WriteString("cannot access the following locations in the snap source directory:\n")
+	for _, p := range e.paths {
+		b.WriteString("- ")
+		b.WriteString(p)
+		b.WriteString("\n")
+	}
+	if len(e.paths) == maxErrPaths {
+		fmt.Fprintf(&b, "- too many errors, listing first %v entries\n", maxErrPaths)
+	}
+	return b.String()
+}
+
+func verifyContentAccessWalkDir(source, prefix string) error {
+	dir, err := os.Open(source)
+	if err != nil {
+		return err
+	}
+	defer dir.Close()
+
+	var errPaths errPathsNotReadable
```

Comment:
I wonder if instead of creating a list of bad paths in each call of `verifyContentAccessWalkDir` we could get it as a parameter and fill it directly. Then we wouldn't need to merge lists later because we would be already adding to the same list.